miner/worker.go

```diff
@@ -287,6 +294,27 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
 	if tx.Type() == types.BlobTxType {
 		return miner.commitBlobTransaction(env, tx)
 	}
+
+	// If a conditional is set, check prior to applying
+	if conditional := tx.Conditional(); conditional != nil {
+		now, cost := time.Now(), conditional.Cost()
+		if !miner.conditionalLimiter.AllowN(now, cost) {
+			return fmt.Errorf("exceeded rate limit: cost %d, tokens %f: %w", cost, miner.conditionalLimiter.Tokens(), errTxConditionalRateLimited)
+		}
+
+		// ditch the reservation as we've checked that we're allowed `cost` units by the limiter
+		miner.conditionalLimiter.ReserveN(now, cost)
```

Comment:
I'm not sure this looks right. You are checking if there is enough tokens to proceed immediately, but erroring right away. Then you reserve them, presumably there's a race here.

I think you want to take the ReserveN, check if it's available immediately, and if it isn't, cancel the reservation so someone else can have it.

Reply:
the miner runs in a single threaded loop so this should eliminate the race condition right?

But ah, didn't see that a reservation could be canceled, that actually feels cleaner! When implementing it looked like the holder would have to block once taking the rez. Thanks for pointing that out

Reply:
Not a data race. Consider if multiple requests passed the Allow check, and then all dropped the Reserve. What would happen is a huge burst that would disallow any new requests for quite sometime.

If it behaved as you described, after they all passed the allow, some of them would actually block, which also not what I think you intended.

I still don't think it's right, the rate package API is pretty confusing:

I think you want to take the reservation. Checking it's OK only tells you if it's possible to satisfy the token count you requested before the end of the universe (1<<63 nano seconds after 1970). You still need to check what the delay is, so check for `reservation.Delay() > 0`. If it's true, cancel your reservation (again assuming you don't want to block...), `reservation.Cancel()`. Anyone that gets `reservation.Delay==0` is ready to proceed immediately.

Reply:
ahh yea good catch, thanks for the explainer

Reply:
> I also spotted that the rate Limiter parameters were passed in the wrong order.

Hmm, I intentionally had it that order but maybe misunderstanding the API. My thinking was that the limiter is refilled at a fixed rate 1000 units per second. And we allow the bucket size to be configurable. So for example with a  value of 5000, if 1000 units was consumed with each conditional, this configuration would only allow 5 conditional txs in a given second.

With the arguments reversed, only 1000 units could ever be consumed in a given second. This is understanding correct?

Reply:
Ah, I'm bummed you didn't just take the commit.

You currently have it so that 1000 units are available per second, with a "debt" (burst) of `RollupTransactionConditionalBurstRate`. Incidentally you actually have `RollupTransactionConditionalBurstRate` set to 1000 too, FWIW.

Reply:
ohh wait i see. In order for the burst to match the limit, we should instead set them to be equal to each other? I think it's simpler to limit the burst to whatever the rate limit is set to